### PR TITLE
Add source-build support.

### PR DESCRIPTION
**Pull request by tmds**
Contributes to https://github.com/dotnet/source-build/issues/1215.

@directhex @MichaelSimons @premun ptal

cc @omajid @ashnaga

### CODE REVIEW

**Comment by tmds** [comment]
The changes to `llvm.proj` are for supporting non-crossbuilds.

There are a few issues when the vmr builds this through arcade:

1.`_InitializeAssemblyVersion` which is used in `packages.builds` is not found. This target should come from `Microsoft.DotNet.Arcade.Sdk`:

```
    /home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/src/nuget/packages.builds(38,11): error MSB4057: The target "_InitializeAssemblyVersion" does not exist in the project.
  ##vso[task.logissue type=error;sourcepath=/home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/src/nuget/packages.builds;linenumber=38;columnnumber=11;code=MSB4057;](NETCORE_ENGINEERING_TELEMETRY=Build) The target "_InitializeAssemblyVersion" does not exist in the project.
```

2.  ~I copied the LICENSE file to the project root to silence `Microsoft.DotNet.Arcade.Sdk/tools/SourceBuild/AfterSourceBuild.proj` which expects to find the license there:~

edit: no longer an issue since the branch was changed, and the new target branch already includes a LICENSE file in the root.

3. The first time this builds, the `microsoft.dotnet.build.tasks.packaging` fails to find the `Icon.png`. Starting the build a second time, the issue disappears. This is the error message (it appears for several `pkgproj`, the output only shows one):

```
  /home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/package-cache/microsoft.dotnet.build.tasks.packaging/8.0.0-beta.23316.6/build/Packaging.targets(1264,5): error : Error when creating nuget lib package from /home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/src/artifacts/packages/Release/specs/runtime.linux-x64.Microsoft.NETCore.Runtime.JIT.Tools.nuspec. NuGet.Packaging.Core.PackagingException: The icon file 'Icon.png' does not exist in the package. [/home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/src/nuget/Microsoft.NETCore.Runtime.JIT.Tools/Microsoft.NETCore.Runtime.JIT.Tools.pkgproj]
  /home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/package-cache/microsoft.dotnet.build.tasks.packaging/8.0.0-beta.23316.6/build/Packaging.targets(1264,5): error :    at NuGet.Packaging.PackageBuilder.ValidateIconFile(IEnumerable`1 files, String iconPath) [/home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/src/nuget/Microsoft.NETCore.Runtime.JIT.Tools/Microsoft.NETCore.Runtime.JIT.Tools.pkgproj]
  /home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/package-cache/microsoft.dotnet.build.tasks.packaging/8.0.0-beta.23316.6/build/Packaging.targets(1264,5): error :    at NuGet.Packaging.PackageBuilder.Save(Stream stream) [/home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/src/nuget/Microsoft.NETCore.Runtime.JIT.Tools/Microsoft.NETCore.Runtime.JIT.Tools.pkgproj]
  /home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/package-cache/microsoft.dotnet.build.tasks.packaging/8.0.0-beta.23316.6/build/Packaging.targets(1264,5): error :    at Microsoft.DotNet.Build.Tasks.Packaging.NuGetPack.Pack(String nuspecPath, String nupkgPath, Manifest manifest, Boolean packSymbols) in /_/src/arcade/artifacts/source-build/self/src/src/Microsoft.DotNet.Build.Tasks.Packaging/src/NuGetPack.cs:line 304 [/home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/src/nuget/Microsoft.NETCore.Runtime.JIT.Tools/Microsoft.NETCore.Runtime.JIT.Tools.pkgproj]
  /home/tmds/repos/dotnet/src/llvm-project/artifacts/source-build/self/package-
```

@MichaelSimons @premun can you provide some assistance with these issues?

**Comment by MichaelSimons** [comment]
It looks like this repo also needs a source-build CI enabled.

**Comment by MichaelSimons** [comment]
> @MichaelSimons @premun can you provide some assistance with these issues?

I can take a look but it won't be until later in the week.  I am fully booked today.

**Comment by tmds** [comment]
> I can take a look but it won't be until later in the week. I am fully booked today.

@MichaelSimons that would be great. I'm sure you'll make faster progress debugging this than I would.

@directhex can you please review the changes I've made to `llvm.proj`?

File: LICENSE.TXT
Comment (akoeplinger) [inline LICENSE.TXT]: instead of copying the license could you add this to the sourcebuild props:
`<PackageLicenseFile>$(RepoRoot)/llvm/LICENSE.TXT</PackageLicenseFile>`

though like I mentioned in https://github.com/dotnet/llvm-project/pull/242#discussion_r1102876858 I'm still curious why we need that.

Comment (tmds) [reply LICENSE.TXT]: I had tried that, but Arcade didn't like it.

From the errors I've seen, I think this gets used by some some intermediate packages that Arcade creates.

Comment (akoeplinger) [reply LICENSE.TXT]: Hmm okay, thanks for trying. It sounds like something that should be fixed so I'd prefer having an issue tracking that.

Comment (tmds) [reply LICENSE.TXT]: I hope @MichaelSimons takes a look as part of the second item on https://github.com/dotnet/llvm-project/pull/439#issuecomment-1598586327.
If we can't fix it for this PR (maybe I was doing it wrong), we should create an issue.

**Comment by tmds** [comment]
Did I break the Windows arm64 builds?

```
llvm.proj(139,5): error MSB3073: (NETCORE_ENGINEERING_TELEMETRY=Build) The command "
    :: VisualStudio includes vswhere.exe that can be used to locate current VisualStudio installation.
    set VSWHERE_TOOLS_BIN=%ProgramFiles(x86)%\Microsoft Visual Studio\Installer\vswhere.exe
    set VS_VCINSTALL_DIR=

    :: Try to locate installed VisualStudio VC environment.
    if "%VCINSTALLDIR%" == "" if exist "%VSWHERE_TOOLS_BIN%" (
        for /f "tokens=*" %%a in ('"%VSWHERE_TOOLS_BIN%" -latest -prerelease -property installationPath') do (
            set VS_VCINSTALL_DIR=%%a\VC\
        )
    )

    if NOT "%VCINSTALLDIR%" == "" set VS_VCINSTALL_DIR=%VCINSTALLDIR%

    :: Run VS build environment script.
    call "%VS_VCINSTALL_DIR%\Auxiliary\Build\vcvarsamd64_arm64.bat"

     ninja " exited with code 1.
```

**Comment by directhex** [comment]
> Did I break the Windows arm64 builds?

Yes.

The Windows build is split into two parts:

* Compile various build-time tools needed by LLVM for $HOST
* Execute those tools whilst building for $TARGET

The error you're getting is:

`This version of D:\a\_work\1\s\artifacts\obj\BuildRoot-arm64\bin\clang-ast-dump.exe is not compatible with the version of Windows you're running. Check your computer's system information and then contact the software publisher.` which implies it's trying to build/use ARM64 `clang-ast-dump.exe` on x64 build VMs.

**Comment by directhex** [comment]
The core issue is for Windows, the host/target steps are separated out in azure-pipelines.yml:

```
        - powershell: eng\build.ps1 -ci -restore -build -arch x64 -configuration $(_BuildConfig) $(_InternalBuildArgs) /p:BuildLLVMTableGenOnly=true
          displayName: 'Build host llvm-tblgen for cross-compiling'

        - powershell: eng\build.ps1 -ci -restore -build -pack $(archflag) -configuration $(_BuildConfig) $(_InternalBuildArgs) $(LLVMTableGenArg) $(ClangTableGenArg)
          displayName: 'Build and package'
```

We'll need to translate this logic to MSBuild for source-build to go through in a single pass.

**Comment by tmds** [comment]
@directhex I've fixed the Windows arm64 CI build issue by making it cross-build aware. CI is passing now. Can you review the changes to `llvm.proj`?

**Comment by MichaelSimons** [comment]
A source-build CI leg should be enabled as well.  Instructions can be found at https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build/onboarding/ci-onboarding.md.

**Comment by directhex** [comment]
This will need re-targeting against `dotnet/main-16.x`

**Comment by tmds** [comment]
> A source-build CI leg should be enabled as well. Instructions can be found at https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build/onboarding/ci-onboarding.md.

@MichaelSimons it seems like we're in the _"The simplest way to onboard"._ -case with the `eng/azure-pipelines.yml` file:

https://github.com/dotnet/llvm-project/blob/bd3acea90cae502d9a62164b863708d57d7993bc/eng/azure-pipelines.yml#L21-L25

What `platforms` should I add under `sourceBuildParameters`?

**Comment by tmds** [comment]
> This will need re-targeting against `dotnet/main-16.x`

I've updated the target branch.

**Comment by jkotas** [comment]
> What platforms should I add under sourceBuildParameters?

The list of the platforms in the doc is way out of date. It lists platforms that are no longer supported, it uses stale docker image links, etc.

I think it is sufficient to add a single non-portable build, say Debian 11. It is unlikely that this repo is going to have distro-specific source build issues.

**Comment by tmds** [comment]
> The list of the platforms in the doc is way out of date. It lists platforms that are no longer supported, it uses stale docker image links, etc.

I also don't find any repo that uses `sourceBuildParameters` which I can use as a base.

**Comment by MichaelSimons** [comment]
@NikolaMilosavljevic - Can you help @tmds get the source-build CI leg enabled?  Please make sure there are tracking issues/PRs to ensure the guidelines are up-to-date.  TIA.

**Comment by NikolaMilosavljevic** [comment]
@tmds it is not necessary to specify any platform. Default platform would be used and that should be sufficient for this repo - https://github.com/dotnet/arcade/blob/e4bd767159fde419b50dd54c6d83e1ef970a68d0/eng/common/templates/jobs/source-build.yml#L42-L46

**Comment by NikolaMilosavljevic** [comment]
@tmds the only other repo building non-managed for source-build, is `runtime`. It appears that they use `almalinux-8`: https://github.com/dotnet/runtime/blob/18321c608e32bc5d13ad0299822cf02ab7ebfd6c/eng/pipelines/common/templates/pipeline-with-resources.yml#L62-L64

Here are the job parameters: https://github.com/dotnet/runtime/blob/18321c608e32bc5d13ad0299822cf02ab7ebfd6c/eng/pipelines/runtime-official.yml#L423-L435

Note that `runtime` uses a different job template.

@ViktorHofer does `almalinux-8` make the most sense for `llvm` repo's source-build leg? This would produce packages that will be consumed by other repos.

**Comment by jkotas** [comment]
> does almalinux-8 make the most sense for llvm repo's source-build leg?

almalinux-8 makes sense. (dotnet/runtime uses almalinux-8 as a proxy for Red Hat 8.)

**Comment by tmds** [comment]
@NikolaMilosavljevic it looks like the source build leg gets cancelled after an hour. How can I increase the time-out?

**Comment by NikolaMilosavljevic** [comment]
> @NikolaMilosavljevic it looks like the source build leg gets cancelled after an hour. How can I increase the time-out?

You can set `timeoutInMinutes` similarly to how it's done in `runtime`: https://github.com/dotnet/runtime/blob/a2a1885f8d0f66546788c758c62013c3da10de86/eng/pipelines/runtime-official.yml#L435 This is further consumed here: https://github.com/dotnet/runtime/blob/a2a1885f8d0f66546788c758c62013c3da10de86/eng/pipelines/common/global-build-job.yml#L47

File: eng/common/templates/jobs/source-build.yml
Comment (NikolaMilosavljevic) [inline eng/common/templates/jobs/source-build.yml:1]: This should be reverted. This is only for `managed` pipelines and shouldn't be updated. Also, this file belongs to `arcade` and will be overwritten when repo gets new arcade update.

File: eng/azure-pipelines.yml
Comment (NikolaMilosavljevic) [inline eng/azure-pipelines.yml:30]: This would be the right place to specify source-build platform - right below `enableSourceBuild: true` line:
```
      sourceBuildParameters:
        platforms:
        - name: 'AlmaLinux_8_Portable'
          container: 'mcr.microsoft.com/dotnet-buildtools/prereqs:almalinux-8-source-build'
```

File: eng/common/templates/jobs/source-build.yml
Comment (tmds) [reply eng/common/templates/jobs/source-build.yml:1]: > shouldn't be updated.

I understand.

> This is only for managed pipelines

Unless I misunderstood, I think you suggested I should use this in https://github.com/dotnet/llvm-project/pull/439#issuecomment-1622223774?

I'm a bit at a loss as to what I should be adding instead. I know little about the ci configuration.

**Comment by tmds** [comment]
> You can set timeoutInMinutes similarly to how it's done in runtime: https://github.com/dotnet/runtime/blob/a2a1885f8d0f66546788c758c62013c3da10de86/eng/pipelines/runtime-official.yml#L435 This is further consumed here: https://github.com/dotnet/runtime/blob/a2a1885f8d0f66546788c758c62013c3da10de86/eng/pipelines/common/global-build-job.yml#L47

@NikolaMilosavljevic where should I set this so it affects the source-build leg?

**Comment by NikolaMilosavljevic** [comment]
> > You can set timeoutInMinutes similarly to how it's done in runtime: https://github.com/dotnet/runtime/blob/a2a1885f8d0f66546788c758c62013c3da10de86/eng/pipelines/runtime-official.yml#L435 This is further consumed here: https://github.com/dotnet/runtime/blob/a2a1885f8d0f66546788c758c62013c3da10de86/eng/pipelines/common/global-build-job.yml#L47
> 
> @NikolaMilosavljevic where should I set this so it affects the source-build leg?

I'm not sure as I've never had to configure timeout for the default source-build job. One thing to try, without having to resort to custom source-build job definition, would be to add `timeoutInMinutes` parameter under `sourceBuildParameters`, so it becomes:
```
      sourceBuildParameters:
        platforms:
        - name: 'AlmaLinux_8_Portable'
          container: 'mcr.microsoft.com/dotnet-buildtools/prereqs:almalinux-8-source-build'
        timeoutInMinutes: 90
```

**Comment by mmitche** [comment]
Wow 2hrs (for the other legs) is pretty hefty. @tmds What kind of performance are you seeing on your machine/expecting?

**Comment by tmds** [comment]
> Wow 2hrs (for the other legs) is pretty hefty. @tmds What kind of performance are you seeing on your machine/expecting?

On my laptop, running `./build.sh` for this repo takes about an hour.

Maybe we want to make the source-built aot feature opt-in/out so it can be disabled where it is not needed.

**Comment by tmds** [comment]
> I'm not sure as I've never had to configure timeout for the default source-build job. One thing to try, without having to resort to custom source-build job definition, would be to add timeoutInMinutes parameter under sourceBuildParameters, so it becomes:

I have tried this, but unfortunately it does not work. @NikolaMilosavljevic what should I do now?

**Comment by mmitche** [comment]
@tmds I don't think that parameter is getting correctly handled on the AzDO template side. Fixing.

File: eng/azure-pipelines.yml
Comment (mmitche) [inline eng/azure-pipelines.yml]:
```suggestion
          jobProperties:
            timeoutInMinutes: 600
```

Comment (mmitche) [reply eng/azure-pipelines.yml]: @tmds I think this will do it based on the way the templates are written.

**Comment by tmds** [comment]
The source-build CI leg is working. It is now failing on one of the issues I had mentioned in https://github.com/dotnet/llvm-project/pull/439#issuecomment-1598586327.

@MichaelSimons can you take a look?

**Comment by MichaelSimons** [comment]
@NikolaMilosavljevic - Can you help @tmds out by investigating the source-build failure?  TIA

**Comment by NikolaMilosavljevic** [comment]
Logs are missing - trying to rerun checks.

**Comment by akoeplinger** [comment]
@NikolaMilosavljevic FYI if you hit the `No commit pushedDate` error that's an AzDO bug that can be worked around by retrying the `/azp run` comment until it works 😄

**Comment by akoeplinger** [comment]
looks like it fails here because it can't find `_InitializeAssemblyVersion` from arcade: https://github.com/dotnet/llvm-project/blob/0dc0f720286df7faa7e727c1c3535f79753f10d7/nuget/packages.builds#L32-L44

**Comment by NikolaMilosavljevic** [comment]
> looks like it fails here because it can't find `_InitializeAssemblyVersion` from arcade:
> 
> https://github.com/dotnet/llvm-project/blob/0dc0f720286df7faa7e727c1c3535f79753f10d7/nuget/packages.builds#L32-L44

I'll investigate this. cc @mmitche for Arcade-specific target that's not getting imported.

**Comment by NikolaMilosavljevic** [comment]
@tmds @akoeplinger why source-build leg takes more than 2 hours to build? It was supposed to only build a very small subset of this repo. I presume there is a work to condition projects - it would make sense to complete that work, first, to minimize build issues.

**Comment by akoeplinger** [comment]
I looked into it a bit and it seems to be happening because `_FilteredTargetFrameworks` is empty and sets `ExcludeFromBuild` here: https://github.com/dotnet/arcade/blob/19a7052872ccc1b676191b4cc85954bf61484ef3/src/Microsoft.DotNet.Arcade.Sdk/tools/TargetFrameworkFilters.BeforeCommonTargets.targets#L16

This then in turn sets `_SuppressAllTargets` which disables importing the Versioning.targets from arcade.

We should be able to turn this off by adding `NoTargetFrameworkFiltering=true` to the root Directory.Build.props, like we do in dotnet/runtime: https://github.com/dotnet/runtime/blob/f60757ad40a36b44d471f1cb72614368eb51a1fd/Directory.Build.props#L314-L315

**Comment by jkotas** [comment]
> why source-build leg takes more than 2 hours to build? 

This builds both Native AOT objwriter and Mono LLVM codegen backend.

To enable native AOT in source build, we can get away by source build building objwriter only that should take significantly less time. It is going to be kicking the can down the road. It is likely that we will want to support building the Mono LLVM codegen backend in the fullness of time.

**Review by akoeplinger — APPROVED** [review]

**Comment by akoeplinger** [comment]
The source-build leg is green 🎉 

@NikolaMilosavljevic please check whether the changes I pushed look good, then this should be ready to go.

**Comment by MichaelSimons** [comment]
> This builds both Native AOT objwriter and Mono LLVM codegen backend.
> 
> To enable native AOT in source build, we can get away by source build building objwriter only that should take significantly less time. It is going to be kicking the can down the road. It is likely that we will want to support building the Mono LLVM codegen backend in the fullness of time.

Adding LLVM into source-build is going to have a significant impact on dependency flows specifically at sdk->installer.  Adding 2+ hours to source-build is going to cause PR validation to rarely finish before another update flows in therefore likely causing large delays.  Are the ways that we can mitigate this?

I have no objection to merging this PR but it is the corresponding runtime change that adds a source-build dependency in Version.Details.xml which would trigger LLVM to be included in the VMR that we should hold off on until we have a plan that would not block the dependency flow.

cc @mmitche

**Comment by akoeplinger** [comment]
We can pass `/p:BuildObjWriterOnly=true` in the source-build leg so it only builds objwriter for now to avoid that but as @jkotas said we'll probably have to build the whole thing at some point.

**Comment by MichaelSimons** [comment]
> We can pass `/p:BuildObjWriterOnly=true` in the source-build leg so it only builds objwriter for now to avoid that but as @jkotas said we'll probably have to build the whole thing at some point.

I would be fine with that.  At the point when we would need to source-build the whole thing, we would need mitigations/features in place to limit the impact to source-build PR validation.

**Comment by mmitche** [comment]
> Adding LLVM into source-build is going to have a significant impact on dependency flows specifically at sdk->installer. Adding 2+ hours to source-build is going to cause PR validation to rarely finish before another update flows in therefore likely causing large delays. Are the ways that we can mitigate this?

Unless the PR _fails_ at any point (a PR check is failing), Maestro will not add new commits on top of a PR that is currently in validation. So it shouldn't cycle endlessly in the best case. But, failures are common so you can get into a bad spot.

I think overall that only enabling the objwriter is the right thing to do for .NET 8.

**Comment by NikolaMilosavljevic** [comment]
> The source-build leg is green 🎉
> 
> @NikolaMilosavljevic please check whether the changes I pushed look good, then this should be ready to go.

They look good. I'll sign-off after source-build leg is scoped down to build just the ObjWriter.

**Comment by Thefrank** [comment]
From a personal experience of working with this repo on FreeBSD, the ObjWriter takes minutes (15 at most) to build whereas the entirety of ObjWriter+JITTools+LLVM takes over 2 hours. At 2+ hours this repo alone would add almost 40% to the build time of a source-build (again, for FreeBSD)

**Comment by omajid** [comment]
As a distro maintainer, it also bothers me that we are rebuilding llvm, which many distros already maintain and package separately. Is there no way we can just build the custom subcomponents like ObjWriter but re-use the llvm packages that distros already maintain? Is it possible to upstream the changes in our llvm fork?

**Comment by directhex** [comment]
LLVM upstream is an absolute nightmare to work with. Personally, I gave up sending critical Windows build system fixes – they have no reviewers, and are deeply committed to the world’s worst patch submission system (and yes I’m including .diff by email in that comparison). That’s just issues orthogonal to “will our changes actually be considered upstreamable by upstream)
Mono absolutely won’t work with an upstream LLVM, we need our various changes for it to work. I’m not certain for objwriter, but I’m reasonably sure it also won’t work with upstream LLVM (or upstream-with-objwriter-itself-patched-in-since-its-original-code-by-us) – changes are needed in various headers to support it.

**Comment by akoeplinger** [comment]
@NikolaMilosavljevic hmm I found no way to pass custom properties into the source build leg via the .yml template so that it ends up being passed to build.sh, is that something we'd need to add in arcade?

**Comment by NikolaMilosavljevic** [comment]
> @NikolaMilosavljevic hmm I found no way to pass custom properties into the source build leg via the .yml template so that it ends up being passed to build.sh, is that something we'd need to add in arcade?

cc @mmitche

**Comment by jkotas** [comment]
In dotnet/runtime, we customize what does and does not get build using the `DotNetBuildFromSource` property: https://github.com/search?q=repo%3Adotnet%2Fruntime%20DotNetBuildFromSource

Can we do the same thing here?

**Comment by mmitche** [comment]
Agreed with @jkotas. There are some other options with `eng/SourceBuild.props` e.g. https://github.com/dotnet/aspnetcore/blob/main/eng/SourceBuild.props if that doesn't work.

**Comment by NikolaMilosavljevic** [comment]
> In dotnet/runtime, we customize what does and does not get build using the `DotNetBuildFromSource` property: https://github.com/search?q=repo%3Adotnet%2Fruntime%20DotNetBuildFromSource
> 
> Can we do the same thing here?

Maybe we could define `DotNetBuildFromSource` in common import, and set it `false`, and only set this property to `true` in projects that we do want to build. This will need to be updated post-8.0, to build everything, but it should unblock the 8.0 work.

**Comment by mmitche** [comment]
> > In dotnet/runtime, we customize what does and does not get build using the `DotNetBuildFromSource` property: https://github.com/search?q=repo%3Adotnet%2Fruntime%20DotNetBuildFromSource
> > Can we do the same thing here?
> 
> Maybe we could define `DotNetBuildFromSource` in common import, and set it `false`, and only set this property to `true` in projects that we do want to build. This will need to be updated post-8.0, to build everything, but it should unblock the 8.0 work.

Do you mean `ExcludeFromSourceBuild`?

**Comment by NikolaMilosavljevic** [comment]
> `ExcludeFromSourceBuild`

Yeah, that is the one that's meant for project-level conditioning. The other property would work as well, but would require more changes.

**Comment by akoeplinger** [comment]
I pushed some changes so we're only building objwriter in the source build job, it completes in ~30 mins now on AzDO.
@NikolaMilosavljevic please take one more look.

**Review by NikolaMilosavljevic — APPROVED** [review]

**Comment by akoeplinger** [comment]
Thanks everyone for the patience! :shipit:

**Comment by MichaelSimons** [comment]
Thanks @akoeplinger for getting this in!

**Pull request merged**
